neat_ml/tests/test_plotting.py

```diff
@@ -0,0 +1,379 @@
+    writer(file_path=csv, output_path=out_png, **extra_kwargs)
+ 
+    result = compare_images(
+        str(baseline_dir / fname), 
+        str(out_png), 
+        tol=2e-2)
+    assert result is None
+
+def test_plot_two_scatter_visual_regression(
+    tmp_path: Path,
+    baseline_dir: Path,
+):
+    csv1 = tmp_path / "scatter_A.csv"
+    pd.DataFrame({"X": [1, 2, 3], "Y": [3, 2, 1]}).to_csv(csv1, index=False)
+
+    csv2 = tmp_path / "scatter_B.csv"
+    pd.DataFrame({"X": [1, 2, 3], "Y": [1, 2, 3]}).to_csv(csv2, index=False)
+
+    out_png = tmp_path / "out" / "plot_two_scatter.png"
+    pmf.plot_two_scatter(
+        csv1_path=csv1,
+        csv2_path=csv2,
+        output_path=out_png,
+        xlim=[0, 4],
+        ylim=[0, 4],
+    )
+    result = compare_images(
+        str(baseline_dir/ "plot_two_scatter.png"), 
+        str(out_png), 
+        tol=2e-2)
+    assert result is None
+
+@pytest.mark.parametrize("json_file, out_dict, err_msg, err_type",
+    [
+        (
+            "bad_params.json",
+            {"MODEL_A": 0.1, "MODEL_B": 0.0},
+            "is missing required keys: {\'MODEL_C\'}",
+            KeyError,
+        ),
+        (
+            "bad_value_type.json",
+            {"MODEL_A": 10.5, "MODEL_B": "not-a-number", "MODEL_C": 20.0},
+            "Parameter 'MODEL_B' must be a number, but got str.",
+            TypeError,
+        ),
+    ]
+)
+def test_load_json_errors(tmp_path: Path,
+json_file, out_dict, err_msg, err_type):
+    """
+    The JSON is valid but omits MODEL_C -> KeyError expected.
+    
+    AND	
+
+    Tests that a TypeError is raised if a parameter value is not a number.
+    
+    The JSON file contains all required keys, but the value for 'MODEL_B'
+    is a string, which should trigger a TypeError.
+    """
+    out_file = tmp_path / json_file
+    json.dump(out_dict, out_file.open("w"))
+    if json_file == "bad_params.json":
+        file_path = os.path.join(tmp_path, json_file)
+        err_msg = f"{file_path} {err_msg}"
+    with pytest.raises(err_type, match=err_msg):
+        pmf.load_parameters_from_json(out_file)
+
+@pytest.mark.parametrize("out_path, out_file, err_msg",
+    [
+        (
+            "empty",
+            None, 
+            "No CSV files found in directory",
+        ),
+        (
+            "bad_csv",
+            pd.DataFrame({"A": [1], "B": [2], "C": [3]}),
+            "CSV",
+        ),
+        (
+            "missing_phase_cols",
+            pd.DataFrame({"A": [1], "B": [2], "C": [3], "D": [4], "E": [5]}),
+            "Dataframe missing phase columns. Skipping.",
+        )
+    ],
+)
+def test_make_phase_diagram_errors(tmp_path: Path,
+out_path, out_file, err_msg):
+    """
+    Empty directory should raise 'No CSV files found in directory'.
+    
+    AND     
+
+    A CSV with only 3 columns triggers the ValueError in the try/except block.
+    
+    AND
+
+    A CSV that does not contain the necessary ``Phase`` columns for plotting
```

Comment:
here and elsewhere, the test docstring explanations probably make more sense inline above their parametrized cases